crypto/src/key/hdkd/mod.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+pub(crate) mod derivation_path;
+
+use crate::key::hdkd::derivation_path::{ChildNumber, DerivationPath};
+
+#[derive(thiserror::Error, Debug, PartialEq, Eq)]
+pub enum DerivationError {
+    #[error("Invalid child number {0}")]
+    InvalidChildNumber(u32),
+    #[error("Malformed child number format")]
+    InvalidChildNumberFormat,
+    #[error("Malformed derivation path format")]
+    InvalidDerivationPathFormat,
+    #[error("Unsupported derivation type")]
+    UnsupportedDerivationType,
+}
+
+pub trait Derivable: Sized + Clone {
+    /// Derive a child private key given a derivation path
+    fn derive_path(&self, path: &DerivationPath) -> Result<Self, DerivationError> {
+        let key = path.into_iter().try_fold(self.clone(), |key, num| key.derive_child(num))?;
+        Ok(key)
```

Comment:
Thanks for that, maybe it was a remnant from a previous version.

Regarding `self` vs `&self`, what are the pros and cons of each approach?

Reply:
> Regarding `self` vs `&self`, what are the pros and cons of each approach?

The tradeoffs here may be quite subtle. With `&self` you avoid some cloning/copying but you `.clone()` explicitly here so this advantage is not as pronounced in this case. My motivation for using `self` here is to avoid the initial `.clone()` and also that the previous iteration produces `self` which can be fed into the next iteration so you already have a value there. Also it makes your trait more general. Since you don't require `T: Clone`, it is in principle implementable for more types. I am not sure if that is useful since keys are just byte arrays that tend to be trivially clonable. However, I can imagine a hypothetical scenario where we deliberately avoid implementing `Clone` on some key-like type to let Rust help us to avoid a key reuse when it is undesirable.

Reply:
Usually in rust, you're explicitly cloning, in the sense that you force the user of the library to clone in the implementation, then that option can be given to the user. Using self here is a good example of how that works. If the user wants to clone, they can clone the outer object instead of consuming this one.